In [8]:
import csv
with open('guns.csv') as file:
    reader = csv.reader(file)
    data = list(reader)
    print(data[:5])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education'], ['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]


In [9]:
headers = data[0]
new_data = data[1:len(data)]
print(headers)
print(new_data[:5])

['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']
[['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', '2']]


In [10]:
years = [row[1] for row in new_data]
year_counts = {}
for year in years:
    if year in year_counts:
        year_counts[year] += 1
    else:
        year_counts[year] = 1
        
print(year_counts)

{'2014': 33599, '2012': 33563, '2013': 33636}


In [11]:
import datetime as dt
dates = [ dt.datetime(year = int(row[1]), month = int(row[2]), day = 1) for row in new_data]
print(dates[:5])

[datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0)]


In [12]:
date_counts = {}
for date in dates:
    if date in date_counts:
        date_counts[date] += 1
    else:
        date_counts[date] = 1

print(date_counts)

{datetime.datetime(2012, 10, 1, 0, 0): 2733, datetime.datetime(2012, 6, 1, 0, 0): 2826, datetime.datetime(2014, 6, 1, 0, 0): 2931, datetime.datetime(2012, 5, 1, 0, 0): 2999, datetime.datetime(2013, 1, 1, 0, 0): 2864, datetime.datetime(2012, 3, 1, 0, 0): 2743, datetime.datetime(2014, 10, 1, 0, 0): 2865, datetime.datetime(2012, 7, 1, 0, 0): 3026, datetime.datetime(2013, 7, 1, 0, 0): 3079, datetime.datetime(2013, 3, 1, 0, 0): 2862, datetime.datetime(2012, 11, 1, 0, 0): 2729, datetime.datetime(2014, 7, 1, 0, 0): 2884, datetime.datetime(2012, 12, 1, 0, 0): 2791, datetime.datetime(2014, 3, 1, 0, 0): 2684, datetime.datetime(2013, 12, 1, 0, 0): 2765, datetime.datetime(2014, 5, 1, 0, 0): 2864, datetime.datetime(2014, 12, 1, 0, 0): 2857, datetime.datetime(2013, 10, 1, 0, 0): 2808, datetime.datetime(2013, 4, 1, 0, 0): 2798, datetime.datetime(2013, 6, 1, 0, 0): 2920, datetime.datetime(2013, 9, 1, 0, 0): 2742, datetime.datetime(2012, 8, 1, 0, 0): 2954, datetime.datetime(2013, 2, 1, 0, 0): 2375, dat

In [13]:
sexes = [row[5] for row in new_data]
races = [row[7] for row in new_data]

sex_counts = {}
for sex in sexes:
    if sex in sex_counts:
        sex_counts[sex] += 1
    else:
        sex_counts[sex] = 1
        
race_counts = {}
for race in races:
    if race in race_counts:
        race_counts[race] += 1
    else:
        race_counts[race] = 1
        
print(sex_counts)
print(race_counts)


{'M': 86349, 'F': 14449}
{'Black': 23296, 'White': 66237, 'Native American/Native Alaskan': 917, 'Asian/Pacific Islander': 1326, 'Hispanic': 9022}


### We analysed the gun deaths against the sex and race. 
### We also, analysed the deaths per year and per month per year. It was noticed that the gun deaths haven't changed much since 2012 and 2014. But we observed that the number of males killed/ died were substantially higher comapred to females.
### Also, the most targeted race is white as ~66000 people were killed whereas the Native Alaskans had the least number of deaths with 917. 
### Furthermore, We can analyse if the victim had a hispanic connection, where majority of the shootings took place and what was the educational background of the victim. 

In [14]:
import csv
with open('census.csv') as file:
    reader = csv.reader(file)
    census = list(reader)
print(census[0:5])

[['Id', 'Year', 'Id', 'Sex', 'Id', 'Hispanic Origin', 'Id', 'Id2', 'Geography', 'Total', 'Race Alone - White', 'Race Alone - Hispanic', 'Race Alone - Black or African American', 'Race Alone - American Indian and Alaska Native', 'Race Alone - Asian', 'Race Alone - Native Hawaiian and Other Pacific Islander', 'Two or More Races'], ['cen42010', 'April 1, 2010 Census', 'totsex', 'Both Sexes', 'tothisp', 'Total', '0100000US', '', 'United States', '308745538', '197318956', '44618105', '40250635', '3739506', '15159516', '674625', '6984195']]


In [16]:
# we first create a dictionary that contains total population for each race 
mapping = {'Asian/Pacific Islander': 15834141, 'Black': 40250635, 'Hispanic': 44618105, 'Native American/Native Alaskan': 3739506, 'White': 197318956}

# we stored the death per race in the following dictionary
race_per_hundredk = {}

# we loop through race_counts to get the killings data for every race 
for race in race_counts:
    total_deaths_per_race = (race_counts[race] / mapping[race]) * 100000
    race_per_hundredk[race] = total_deaths_per_race
    
print(race_per_hundredk)
    

{'Black': 57.8773477735196, 'White': 33.56849303419181, 'Asian/Pacific Islander': 8.374309664161762, 'Native American/Native Alaskan': 24.521955573811088, 'Hispanic': 20.220491210910907}


In [17]:
'''Exploring Homicide rates based on race'''
intents = [row[3] for row in new_data]
races = [row[7] for row in new_data]

homicide_race_per_hundredk = {}

for i, race in enumerate(races):
    if intents[i] == 'Homicide':
        if race in homicide_race_per_hundredk:
            homicide_race_per_hundredk[race] += 1
        else:
            homicide_race_per_hundredk[race] = 1
            
for homicide in homicide_race_per_hundredk:
    total_death_by_homicide = (homicide_race_per_hundredk[homicide] / mapping[homicide]) * 100000
    homicide_race_per_hundredk[homicide] = total_death_by_homicide
    
print(homicide_race_per_hundredk)

{'Black': 48.471284987180944, 'White': 4.6356417981453335, 'Native American/Native Alaskan': 8.717729026240365, 'Asian/Pacific Islander': 3.530346230970155, 'Hispanic': 12.627161104219914}


### We analysed the death that took palce from 2012 to 2014 based on every race 
### We then converted the deaths to 'out of 100,000' to determine the number of deaths for each race per 100,000. To achieve this we divided the deaths every race by the their total population. We gert the total population for every race from the census.csv file. 
### We also find out the number of homicides taking place in each race.
### We can still perform more analyses: 
- Geographically and education-wise to find out more about the background of the victims.
- Figure out links, if any, between month and homicide rate.
- Explore the homicide rate by gender.
- Explore the rates of otehr intents, like Accidental, by gender and race.
- Find out if gun death rates correlate to location and education

In [30]:
"""Exploring homicide rates based on gender"""
gender = [row[5] for row in new_data]
total_population = [row[9] for row in census]
intents = [row[3] for row in new_data]

homicide_gender_per_hundredk = {}

for i, sex in enumerate(gender):
    if intents[i] == 'Homicide':
        if sex in homicide_gender_per_hundredk:
            homicide_gender_per_hundredk[sex] += 1
        else:
            homicide_gender_per_hundredk[sex] = 1
    
print(homicide_gender_per_hundredk)    
for homicide in homicide_gender_per_hundredk:
    total_deaths = (homicide_gender_per_hundredk[homicide] / int(total_population[1])) * 100000
    homicide_gender_per_hundredk[homicide] = total_deaths
    
print(homicide_gender_per_hundredk)

{'M': 29803, 'F': 5373}
{'M': 9.652933024735729, 'F': 1.7402680650238256}


In [28]:
''' Exploring the rate of accidental intents by gender and race'''
gender = [row[5] for row in new_data]
intents = [row[3] for row in new_data]

accidental_by_gender = {}
for i, sex in enumerate(gender):
    if intents[i] == 'Accidental':
        if sex in accidental_by_gender:
            accidental_by_gender[sex] += 1
        else:
            accidental_by_gender[sex] = 1
            
print(accidental_by_gender) 

races = [row[7] for row in new_data]

accidental_race_per_hundredk = {}
for i, race in enumerate(races):
    if intents[i] == 'Accidental':
        if race in accidental_race_per_hundredk:
            accidental_race_per_hundredk[race] += 1
        else:
            accidental_race_per_hundredk[race] = 1
            
for race in accidental_race_per_hundredk:
    total_accidents = (accidental_race_per_hundredk[race] / mapping[race]) * 100000
    accidental_race_per_hundredk[race] = total_accidents
    
print(accidental_race_per_hundredk)

{'M': 1421, 'F': 218}
{'Native American/Native Alaskan': 0.5883130017708221, 'Black': 0.814893976206835, 'White': 0.5736904466492313, 'Hispanic': 0.3249801846133985, 'Asian/Pacific Islander': 0.07578560782046845}


In [32]:
''' Exploring where the shooting happened and the educational background of the victim'''
places = [row[9] for row in new_data]
educational_details = [row[10] for row in new_data]

shooting_places = {}
for place in places:
    if place in shooting_places:
        shooting_places[place] += 1
    else:
        shooting_places[place] = 1
        
print(shooting_places)

education = {}
for edu in educational_details:
    if edu in education:
        education[edu] += 1
    else:
        education[edu] = 1
        
print(education)

{'NA': 1384, 'School/instiution': 671, 'Sports': 128, 'Residential institution': 203, 'Other specified': 13751, 'Trade/service area': 3439, 'Other unspecified': 8867, 'Farm': 470, 'Home': 60486, 'Industrial/construction': 248, 'Street': 11151}
{'1': 21823, 'NA': 53, '3': 21680, '5': 1369, '4': 12946, '2': 42927}


In [39]:
def max_min_killing_place(d):
    return max(d, key = d.get), d[max(d, key = d.get)], min(d, key = d.get), d[min(d, key = d.get)]

def max_min_killing_education(d):
    return max(d, key = d.get), d[max(d, key = d.get)], min(d, key = d.get), d[min(d, key = d.get)]
print(max_min_killing_place(shooting_places))
print(max_min_killing_education(education))

('Home', 60486, 'Sports', 128)
('2', 42927, 'NA', 53)


### From the analysis above we can see that the maximum killings have happened at homes and the least have happened at sports arena from 2012 to 2014.

### Crimes are targeted at people who have graduated from high school or equivalent with 42,927 being killed. There are 53 people killed whose educational background is not available. 

### We can further categorize the killings in every place based on:
- Race
- month
- intent
- age

### We can do a similar analysis for educational background as well.

In [36]:
''' Let's find out the killings that took place in different places based on race'''
races = [row[7] for row in new_data]
places = [row[9] for row in new_data]

place_race_killing = {}
for i, place in enumerate(places):
    if races[i] == 'Black':
        if place in place_race_killing:
            place_race_killing[place] += 1
        else:
            place_race_killing[place] = 1
            
print(place_race_killing)

{'NA': 353, 'School/instiution': 233, 'Residential institution': 43, 'Other specified': 3663, 'Trade/service area': 1023, 'Other unspecified': 2590, 'Industrial/construction': 59, 'Home': 8942, 'Sports': 27, 'Farm': 13, 'Street': 6350}


In [37]:
def race_killing(d):
    return max(d, key = d.get), d[max(d, key = d.get)], min(d, key = d.get), d[min(d, key = d.get)]

print(race_killing(place_race_killing))

('Home', 8942, 'Farm', 13)


### So, we can see that most black people were killed at home (8942) whereas the least number of black people were killed in farms (13).